modules/reactor/src/meshgenerators/SubdomainExtraElementIDGenerator.C

```diff
@@ -0,0 +1,111 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SubdomainExtraElementIDGenerator.h"
+
+registerMooseObject("ReactorApp", SubdomainExtraElementIDGenerator);
+
+#include "libmesh/elem.h"
+
+InputParameters
+SubdomainExtraElementIDGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+  params.addRequiredParam<MeshGeneratorName>(
+      "input", "Name of an existing mesh generator to which we assign element IDs");
+  params.addRequiredParam<std::vector<SubdomainName>>("subdomains",
+                                                      "Subdomain names present in the input mesh");
+  params.addRequiredParam<std::vector<std::string>>("extra_element_id_names",
+                                                    "List of user-defined extra element ID names");
+  params.addRequiredParam<std::vector<std::vector<dof_id_type>>>(
+      "extra_element_ids",
+      "User-defined extra element IDs corresponding to 'subdomains' in the same order");
+
+  params.addParam<std::vector<dof_id_type>>(
+      "default_extra_element_ids", "Default extra element IDs for elements not in 'subdomains'");
+
+  params.addClassDescription(
+      "Assign extra element IDs for elements on a mesh based on mesh subdomains.");
+  return params;
+}
+
+SubdomainExtraElementIDGenerator::SubdomainExtraElementIDGenerator(const InputParameters & params)
+  : MeshGenerator(params),
+    _input(getMesh("input")),
+    _subdomain_names(getParam<std::vector<SubdomainName>>("subdomains"))
+{
+  if (_subdomain_names.size() == 0)
+    paramError("subdomains", "Empty subdomain vector provided!");
+}
+
+std::unique_ptr<MeshBase>
+SubdomainExtraElementIDGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  // construct a map from the subdomain ID to the index in 'subdomain'
+  std::map<SubdomainID, unsigned int> subdomains;
+  for (unsigned int i = 0; i < _subdomain_names.size(); ++i)
+  {
+    SubdomainID id = Moose::INVALID_BLOCK_ID;
+    std::istringstream ss(_subdomain_names[i]);
+
+    if (!(ss >> id) || !ss.eof())
+      id = mesh->get_id_by_name(_subdomain_names[i]);
```

Comment:
Logan added a getSubdomainID to MooseMeshUtils since then. Could you please use that